### PR TITLE
detection_range_model function

### PR DESCRIPTION
**Pull request by jdpye**
In GitLab by @benjaminhlina on Nov 26, 2021, 11:30

detection_range_model uses a third-order polynomial, logit, or, probit model to determine detection range based on a expected percentage of detections for acoustic telemetry receivers. This function is to be used after a preliminary range test which uses multiple distances away from a representative receiver. Preliminary detection efficiency is to be determined in Vemco's Range Testing software and exported as a csv. Vignette on the use of the function is being written.

### CODE REVIEW

**Comment by jdpye** [comment]
@benjaminhlina is looking to revamp this a bit before he officially submits it to glatos.

**Comment by benjaminhlina** [comment]
Hey Jon can you reopen this? I have finished the vignette!! :)

**Comment by mhpob** [comment]
Since mapview is only used in the vignette, it should probably go in Suggests rather than Imports to reduce dependencies (see commit d7d7a83).

**Comment by mhpob** [comment]
@benjaminhlina FYI, mapview is still in Imports.

**Comment by benjaminhlina** [comment]
Thanks Mike, thought I had addressed that but I guess it slipped. I have no removed it from imports.

**Comment by benjaminhlina** [comment]
I have updated this branch - specifically have fixed spelling and grammatical errors in the the vignette. This now should be good to go :) @chrisholbrook

**Comment by chrisholbrook** [comment]
@benjaminhlina @jdpye 
I'm starting review on this now; likely will complete tomorrow morning. (really sorry to have dropped the ball on this but will make effort to get it into dev tomorrow; then dev--> main later this week).

**Comment by benjaminhlina** [comment]
@chrisholbrook Thanks and it's all good - was making the presentation the workshop next week and came upon several spelling/grammar mistakes that have now been corrected.

**Comment by mhpob** [comment]
@benjaminhlina just saw this since I was tagged before. 

Is there a reason you removed the R CMD check action? Also, it doesn't look like the recent several-month's changes in the dev branch were pulled in -h may be causing the merge conflicts that are being warned about.

**Comment by chrisholbrook** [comment]
> Also, it doesn't look like the recent several-month's changes in the dev branch were pulled in -- may be causing the merge conflicts that are being warned about.

I've merged this with dev locally and am resolving the conflicts. @mhpob For feature branches like this, do you suggest the feature branch author merge dev --> feature-branch before submitting a PR to merge feature-branch --> dev? Makes sense to me but I have limited experience with these workflows.

**Comment by mhpob** [comment]
I think it takes work off the package maintainer and lowers the bar to getting the PR accepted. The contributing guidelines ask the feature branch to be sourced from dev, but not that dev changes are merged back in before submitting, so perhaps it's not fair to ask at this time.

**Comment by chrisholbrook** [comment]
> I think it takes work off the package maintainer and lowers the bar to getting the PR accepted. The contributing guidelines ask the feature branch to be sourced from dev, but not that dev changes are merged back in before submitting, so perhaps it's not fair to ask at this time.

It does seem like preferred workflow for everyone if the submitter of the PR makes sure dev is not ahead of feature branch, but after the PR is submitted, it's somewhat out of their control. In this case, we added a bunch to dev after the PR was originally submitted, so it's probably most efficient for me to sort out the differences/conflicts than Ben; at least in this case. And this highlights the problem of maintainer sitting on PRs too long (my bad).

**Comment by chrisholbrook** [comment]
@benjaminhlina Did you intend to change 'data_loading_vignette.Rmd'? I see changes to 'data_loading_vignette.R' and 'data_loading_vignette.html' but not the Rmd file, which is the source from which those others are built (I see no reason why the R and html files should be edited directly).

**Comment by benjaminhlina** [comment]
Hi @chrisholbrook and @mhpob - Sorry about all this. To answer both questions 1) No I did not mean to remove R CMD Check yaml and will reinstate it shortly. Did not realize I had done such. 2) I can update the files that are conflicting if you'd like me to @chrisholbrook. I did not intend to modify the R and HTML file for `data_loading_vignette.rmd`.

File: vignettes/detection_efficiency_vignette.Rmd
Comment (chrisholbrook) [inline vignettes/detection_efficiency_vignette.Rmd:493]: > Error: object 'buffer_rings_pts' not found

It's unclear if you intended to use 'buffer_rings_pts' from earlier script (Section 2) or if you meant to use 'redeploy_loc_pts' here; but this needs to be fixed.

**Review by chrisholbrook — CHANGES_REQUESTED** [review]
Several errors in the vignette and warnings were encountered that should be fixed (errors) or explained to the reader (OK for warnings if they are benign). There are also a number of R CMD CHECK errors, which I am am addressing and will push fixes for today.

File: vignettes/detection_efficiency_vignette.Rmd
Comment (chrisholbrook) [inline vignettes/detection_efficiency_vignette.Rmd:5]: I suggest hardcoding the date here to reflect when this document was actually edited. Otherwise, it simply reflects the installation date.

File: vignettes/detection_efficiency_vignette.Rmd
Comment (benjaminhlina) [reply vignettes/detection_efficiency_vignette.Rmd:145]: This is a common warning with `st_cast()` when geometry lengths differ.

File: vignettes/detection_efficiency_vignette.Rmd
Comment (benjaminhlina) [reply vignettes/detection_efficiency_vignette.Rmd:159]: This is a common warning with `st_cast()` when geometry lengths differ.

File: vignettes/detection_efficiency_vignette.Rmd
Comment (benjaminhlina) [reply vignettes/detection_efficiency_vignette.Rmd:504]: I have amended the vignette to reload and create this object

File: vignettes/detection_efficiency_vignette.Rmd
Comment (benjaminhlina) [reply vignettes/detection_efficiency_vignette.Rmd:259]: I have amended the vignette to reload and create this object so that it does not error considering the object cannot be found considering it isn't brought in.

File: vignettes/detection_efficiency_vignette.Rmd
Comment (benjaminhlina) [reply vignettes/detection_efficiency_vignette.Rmd:505]: This is a common warning with `st_cast()` when geometry lengths differ.

File: vignettes/detection_efficiency_vignette.Rmd
Comment (benjaminhlina) [reply vignettes/detection_efficiency_vignette.Rmd:523]: This is a common warning with `st_cast()` when geometry lengths differ.

File: vignettes/detection_efficiency_vignette.Rmd
Comment (benjaminhlina) [reply vignettes/detection_efficiency_vignette.Rmd]: I have added the hard coded the date here

File: vignettes/detection_efficiency_vignette.Rmd
Comment (benjaminhlina) [reply vignettes/detection_efficiency_vignette.Rmd]: I have updated this to `redeploy_loc_pts` as that is what the correct object should be - sorry about that.

File: vignettes/detection_efficiency_vignette.Rmd
Comment (benjaminhlina) [reply vignettes/detection_efficiency_vignette.Rmd:223]: So if this chunk is meant to be as an example of what someone could use, the chunk has `echo = TRUE` and `eval = FALSE` so that when it knits the `rmd` it doesn't actual run but when running each chunk I found it also to produce the error. I can remove this all together if you'd like but originally wanted to demonstrate how to export these objects.

File: vignettes/detection_efficiency_vignette.Rmd
Comment (benjaminhlina) [reply vignettes/detection_efficiency_vignette.Rmd:226]: The same goes for this chunk as well.

File: vignettes/detection_efficiency_vignette.Rmd
Comment (benjaminhlina) [reply vignettes/detection_efficiency_vignette.Rmd:427]: I'm not sure what is causing this warning message as I cannot produce it on my end. I'll dig into it some more

File: vignettes/detection_efficiency_vignette.Rmd
Comment (benjaminhlina) [reply vignettes/detection_efficiency_vignette.Rmd:366]: I have written into the function to warn the user to confirm the model_frame argument is correct for the syntax that the user has supplied the function when it is set to polynomial.

File: vignettes/detection_efficiency_vignette.Rmd
Comment (benjaminhlina) [reply vignettes/detection_efficiency_vignette.Rmd:391]: This warning is produced by `glm()` when `family` is set to `binomial`. It is telling the user that there are `0`s or `1`s present in the data.

File: vignettes/detection_efficiency_vignette.Rmd
Comment (benjaminhlina) [reply vignettes/detection_efficiency_vignette.Rmd:382]: This warning is produced by `glm()` when `family` is set to `binomial`. It is telling the user that there are `0`s or `1`s present in the data.

File: vignettes/detection_efficiency_vignette.Rmd
Comment (benjaminhlina) [reply vignettes/detection_efficiency_vignette.Rmd]: Fixed - sorry about that

File: vignettes/detection_efficiency_vignette.Rmd
Comment (chrisholbrook) [reply vignettes/detection_efficiency_vignette.Rmd:145]: I agree this is lower priority than other warnings I've flagged here (so OK releasing in this form with intent to clean up later), but readers not familiar with sf will appreciate you explaining to them what this means. In this particular case, this warning tells you that you are making an assumption that all geometries have the same attributes--there are consequences if this is not true. If we intend to show/teach best practices here, this should be cleaned up in the code (set attributes constant among geometries; or run the operation separately on each geometry) so that the warning isn't generated.

**Comment by chrisholbrook** [comment]
completed in commit 6525afd732af525de909f5a69350d3efad735bbf

**Review by chrisholbrook — APPROVED** [review]

**Comment by chrisholbrook** [comment]
@benjaminhlina This is now live in 0.8.0. Thanks for the contribution! I took the liberty of making a few edits to code chunks that were throwing warnings and errors (mentioned in review) that I though should be resolved. Feel free to submit another PR if you'd like to correct any of those changes. Also, note some overlap between your vignette and 'detection_range_handout.html'. I will be encouraging you and others to revisit the topic of detection range analysis with the goal of identifying some consensus best practice. Even as separate vignettes, though, these are valuable contributes.

**Comment by benjaminhlina** [comment]
Thanks @chrisholbrook, I have pulled the changes you have made to this that you merged with dev and agree that all the warnings need to be explained to the user. I also a agree that the save steps need to actually save and not error as the end-user will not know why it's erroring and appreciate you amending those steps as well as correcting the warnings.

**Pull request merged**
